### ISSUE TITLE
Metodologia de avaliação do PPO

### ISSUE THREAD

**Issue by lsoaresesilva** [ifpe-cti/modelo-projeto-ppo#9]
Sugiro alinhar com as avaliações da última unidade das disciplinas que são ofertadas no 4 ano/período, por exemplo:

* Testes de SW: avaliar a presença de testes unitários, construção de mocks e execução de integração contínua
* Arquitetura: padrões de projetos adotados e decisões de arquitetura

Mais algumas sugestões de critérios de avaliação:

* Documentos de requisitos (seja utilizando uma metodologia tradicional ou ágil)
* Completude do projeto (o que entregou vs o que prometeu, vide o critério acima)
* Uso de ferramentas de DEV: git, maven, docks, entre outros.
* Avaliação de usabilidade, ux, etc (professores, se manifestem)
* Implantação do projeto (colocar o projeto hospedado em algum servidor, exemplo o google cloud)
* Uso de base de dados e seu respectivo ER

**Comment by lsoaresesilva** [comment]
Mais algumas sugestões de critérios de avaliação:

* Documentos de requisitos (seja utilizando uma metodologia tradicional ou ágil)
* Completude do projeto (o que entregou vs o que prometeu, vide o critério acima)
* Uso de ferramentas de DEV: git, maven, docks, entre outros.
* Avaliação de usabilidade, ux, etc (professores, se manifestem)
* Implantação do projeto (colocar o projeto hospedado em algum servidor, exemplo o google cloud)
* Uso de base de dados e seu respectivo ER

**Comment by fabriciofx** [comment]
@lsoaresesilva **Ótimas** sugestões. Aqui vai mais algumas:

- Impacto de utilização da solução pela sociedade (é alguma aplicação que há algum impacto na sociedade? Vai ajudar a resolver o problema de várias pessoas/sociedade?)
- O potencial de inovação da solução;

**Comment by fabriciofx** [comment]
Outro ponto que temos que discutir é a questão de documentação. Sabemos hoje em dia, até pelo manifesto ágil, que a maior documentação de um projeto é o próprio código. Além disso, a documentação caduca muito rápido. Seria interessante solicitar somente naquilo estritamente necessário e focar mais no código. É ele que agrega valor final ao produto e não a documentação do projeto. WDYT?

**Comment by fabriciofx** [comment]
@lsoaresesilva Essa ideia de colocar o projeto na internet é excelente. Inclusive, já estou me movimentando nesse sentido. Precisamos avaliar as nossas opções:
- Google Cloud Platform (GCP)
- Amazon Web Services (AWS)
- Heroku
- OpenShift
- HelioHost
Mais algum que eu tenha esquecido?

**Comment by elmanorc** [comment]
Acho importante diferenciarmos o que será obrigatório do que será opcional (pontuação bônus, talvez); ter cuidado para a lista de itens obrigatórios não ser muito vasta. Seria interessante definir inicialmente os critérios essenciais.

**Comment by fabriciofx** [comment]
@elmanorc Concordo. Dentre os critérios apontados por @lsoaresesilva e por mim, **os essenciais**, são:

- Completude do projeto (o que entregou vs o que prometeu)
- Impacto de utilização da solução pela sociedade (é alguma aplicação que há algum impacto na sociedade? Vai ajudar a resolver o problema de várias pessoas/sociedade?)
- Arquitetura: padrões de projetos adotados e decisões de arquitetura
- Testes de SW: avaliar a presença de testes unitários, construção de mocks e execução de integração contínua
- Uso de ferramentas de DEV: git, maven, docks, entre outros.
- Uso de base de dados e seu respectivo ER
- Implantação do projeto (colocar o projeto hospedado em algum servidor, exemplo o google cloud)

E os **bônus**:

- O potencial de inovação da solução;
- Documentos de requisitos (seja utilizando uma metodologia tradicional ou ágil)
- Avaliação de usabilidade, ux, etc (professores, se manifestem)

**Comment by elmanorc** [comment]
Um item que acho muito importante é Documentação do Código. Manutenção é o maior custo de um sistema, e ausência ou probreza de documentação é algo crítico que dificulda sua evolução.

**Comment by fabriciofx** [comment]
@elmanorc Desculpe, mas eu discordo. A maior documentação de um projeto de software é o próprio código + os testes (que também é código). Documentação deve ser a *mínima* e mais *essencial* possível. A razão é que o código muda muito, logo a documentação também irá mudar. Manter essas duas coisas sincronizadas só toma tempo do DEV e não agrega valor ao produto. Inclusive, esse é o segundo ponto do [Manifesto Ágil](http://www.manifestoagil.com.br/):
>Software em funcionamento mais que documentação abrangente

**Comment by elmanorc** [comment]
@fabriciofx  Você não entendeu. Não falo de documentação UML, essas coisas; digo documentação no próprio código (javadoc, etc.).

**Comment by fabriciofx** [comment]
@elmanorc Ah, tá... mas também depende. O quê do javadoc deve ser entregue? E o que é o seu "etc"? Mas note também que documentação, mesmo sendo javadoc, agrega menos valor do que o código propriamente dito. Se há alguma dúvida na utilização/entendimento, o problema IMHO é na API e na falta de testes.

**Comment by fabriciofx** [comment]
Outro ponto que também precisamos avaliar: como será feito avaliação/desempenho individual?

**Comment by lsoaresesilva** [comment]
Quando falei de documentação me referi ao uso de no mínimo estórias do usuário e um backlog, caso utilizem um modelo ágil de desenvolvimento. Eu considero isto o mínimo para se ter um planejamento/roadmap do projeto.

**Comment by fabriciofx** [comment]
@lsoaresesilva Concordo. Inclusive, vários alunos do 4o ano integrado comentário as dificuldades em se produzir código e manter documentação sincronizados (diagramas UML e outros). Assim, precisamos definir o que é essencial, do que é um "bônus", por assim dizer.

**Comment by elmanorc** [comment]
Qualidade do código:

Também julgo importante analisarmos o uso não apenas de padrões de projeto, mas também dos fundamentos básicos da OO como Interfaces, tratamento de exceções, encapsulamento, modificadores de acesso coerentes, etc. 

Na disciplina de LPOOED havia feito a lista dos 10 mandamentos para o projeto da 4a unidade:

https://sites.google.com/site/elmano/teaching/ifpe-2016-pooed-integrado-manha
